Snowflake.Data.Tests/UnitTests/SFGCSClientTest.cs

```diff
@@ -0,0 +1,457 @@
+                .Returns(() =>
+                {
+                    var response = new Mock<HttpWebResponse>();
+
+                    if (httpStatusCode == HttpStatusCode.OK)
+                    {
+                        response.Setup(c => c.Headers).Returns(new WebHeaderCollection());
+                        response.Object.Headers.Add("content-length", MockGCSClient.ContentLength.ToString());
+                        response.Object.Headers.Add(SFGCSClient.GCS_METADATA_SFC_DIGEST, MockGCSClient.SFCDigest);
+                    }
+                    else
+                    {
+                        response.SetupGet(c => c.StatusCode)
+                            .Returns(httpStatusCode);
+                        throw new WebException("Mock GCS Error", null, 0, response.Object);
+                    }
+
+                    return response.Object;
+                });
+            _client.SetCustomWebRequest(mockWebRequest.Object);
+
+            // Act
+            FileHeader fileHeader = _client.GetFileHeader(_fileMetadata);
+
+            // Assert
+            AssertForGetFileHeaderTests(expectedResultStatus, fileHeader);
+        }
+
+        [Test]
+        [TestCase(HttpStatusCode.OK, ResultStatus.UPLOADED)]
+        [TestCase(HttpStatusCode.Unauthorized, ResultStatus.RENEW_TOKEN)]
+        [TestCase(HttpStatusCode.Forbidden, ResultStatus.NEED_RETRY)]
+        [TestCase(HttpStatusCode.InternalServerError, ResultStatus.NEED_RETRY)]
+        [TestCase(HttpStatusCode.ServiceUnavailable, ResultStatus.NEED_RETRY)]
+        [TestCase(HttpStatusCode.NotFound, ResultStatus.NOT_FOUND_FILE)]
+        [TestCase(HttpStatusCode.Conflict, ResultStatus.ERROR)] // Any error that isn't the above will return ResultStatus.ERROR
+        public async Task TestGetFileHeaderAsync(HttpStatusCode httpStatusCode, ResultStatus expectedResultStatus)
+        {
+            // Arrange
+            var mockWebRequest = new Mock<WebRequest>();
+            mockWebRequest.Setup(client => client.GetResponseAsync())
+                .Returns(() =>
+                {
+                    var response = new Mock<HttpWebResponse>();
+
+                    if (httpStatusCode == HttpStatusCode.OK)
+                    {
+                        response.Setup(c => c.Headers).Returns(new WebHeaderCollection());
+                        response.Object.Headers.Add("content-length", MockGCSClient.ContentLength.ToString());
+                        response.Object.Headers.Add(SFGCSClient.GCS_METADATA_SFC_DIGEST, MockGCSClient.SFCDigest);
+                    }
+                    else
+                    {
+                        response.SetupGet(c => c.StatusCode)
+                            .Returns(httpStatusCode);
+                        throw new WebException("Mock GCS Error", null, 0, response.Object);
+                    }
+
+                    return Task.FromResult((WebResponse) response.Object);
```

Comment:
this response.Object is pretty much the same in Sync and Async tests and could be extracted

Reply:
Sure, extracted that code into its own method